Make installing openstack/ironic clients optional

hardys opened the pull request:
For RHEL/CentOS this has proven problematic because the RDO repos
aren't always available for a specific release (e.g RHEL7 and
recently CentOS 8 took a while to become available)

We could rely on pip for all platforms, but this can cause
dependency issues with packaged python packages, and typically
can't be used for debugging on a target cluster (whereas a
container image often can).

So this change provides a shell alias and a wrapper script to provide
a backwards compatible interface, but launch the new ironic-client
image instead - testing indicates this is slightly slower but it
should be sufficient for debug usage which is the primary purpose
of the clients in this context.

hardys commented:
/cc @elfosardo

hardys commented:
/cc @maelk @stbenjam

hardys commented:
Marked WIP because I've not yet tested this on a freshly installed system, and also because I wanted to discuss if we should remove the clients in the ubuntu case as well, for consistency?

hardys commented on centos_install_requirements.sh:
```diff
@@ -46,17 +46,6 @@ sudo yum -y install \
-# the version shipped in the tripleo repos. This needs further
-# investigation.
-sudo yum -y update --exclude=oniguruma
-
```
I'm not entirely sure what to do with existing deployments that have RDO repos enabled - at first glance it seems like we'd disable/remove them - but as mentioned here there are actually $many packages that get pulled in from the RDO deps repo, either due to dependencies or because they are newer.

stbenjam replied:
Perhaps we could print a message if we detect the RDO repos are installed that they are no longer required by Metal3, and users can disable them if desired? I don't think I'd want to disable the repos in case the user is using them for something else, but just leaving them enabled without saying anything isn't great either.

hardys replied:
My main concern is more dependency conflicts like the oniguruma one - if we leave the repos around that may happen again, but we can't really fully remove the dependency on the repo on an existing system (on my test machine over 2000 packages are pulled in from the openstack repos!!)

I guess the best we can do is log a warning if we find them, and be sure to not enable them for net-new platforms (e.g CentOS8), then over time we can fully break this dependency

stbenjam commented on 02_configure_host.sh:
```diff
@@ -198,3 +198,15 @@ sudo "${CONTAINER_RUNTIME}" run -d --net host --privileged --name vbmc ${POD_NAM
+    echo "alias openstack=\"${SCRIPTDIR}/openstackclient.sh\"" >> ~/.bashrc 
+    # Set also for the current shell
+    # shellcheck disable=SC2139
+    alias openstack="${SCRIPTDIR}/openstackclient.sh"
```
Do aliases created in a script persist? They don't seem to in my testing

hardys replied:
I thought it did but will check, thanks

hardys replied:
Yeah you're right, it makes sense that we can't set something in the parent shell from the subshell so I must have made a mistake in my testing.

I guess the only option which won't require starting a new shell session is to copy or symlink the wrapper script to somewhere already on the $PATH?

elfosardo replied:
To be able to make that persistent in the current shell, you'd have to source the script and not just run it.
Best way to create a permanent alias would be adding it to .bashrc, as you're doing, but I'm not sure it's a good idea, I'd rather add a message as @stbenjam suggested.
If we want to redirect the openstack command to the openstackclient.sh script inside the scripts, I suggest to use a function or redefine the openstack command as a variable.

hardys replied:
Looking at `01_prepare_host.sh` I see we do use /usr/local/bin by default for kubectl:

```
KUBECTL_PATH="${KUBECTL_PATH:-/usr/local/bin/kubectl}"
```

Perhaps I should do the same and copy the wrapper script to /usr/local/bin/openstack (with the option to override the location e.g OPENSTACK_CLIENT_PATH similar to the above?

That should work in most cases except on some platforms running as root I think?

stbenjam commented on 02_configure_host.sh:
```diff
@@ -198,3 +198,15 @@ sudo "${CONTAINER_RUNTIME}" run -d --net host --privileged --name vbmc ${POD_NAM
+# runs the clients in a container (metal3-io/ironic-client)
+if ! command -v openstack; then
+  if ! grep -q "^alias openstack" ~/.bashrc; then
+    echo "alias openstack=\"${SCRIPTDIR}/openstackclient.sh\"" >> ~/.bashrc 
```
Not everyone uses bash though. Also, if $SCRIPTDIR changes, their alias won't get updated.

I think just printing some message with the path to openstackclient.sh at the very end of metal3-dev-env run would be better.

hardys replied:
Bash is the default shell on many/most platforms these days so I was trying to reduce friction in the common case - I can grep for the entire string though to catch someone changing the $SCRIPTDIR (this seems unlikely but it's a possible corner-case for sure).

We can definitely also print some message and/or document the running the script directly - this was really just a shim to ensure backwards compatibility and avoid breaking existing users who are used to typing `openstack foo`

maelk commented:
That looks good. I would also remove it from Ubuntu. About the bash issue, can it be both in the bashrc and a message be printed so that those not using bash know how to set the alias ? Also putting this in the docs would be helpful with that regard. either somewhere in this repo or in the try-it.md

hardys commented on vm-setup/roles/packages_installation/defaults/main.yml:
```diff
@@ -28,14 +28,11 @@ packages:
      - requests
      - setuptools
      - libvirt-python
-     - python-openstackclient
```
I don't have an ubuntu box to test this on - anyone know if these other deps were added for the clients?  Seems likely but I don't want to remove them without testing :)

hardys replied:
Ok so I did some testing on a freshly installed RHEL8 box, and it seems the lxml, netaddr and libvirt-python dependencies are required for ansible - we can probably remove some of these though (certainly virtualbmc and on RHEL I removed crudini, pip, requests, setuptools and nodejs)

maelk replied:
virtualbmc can be removed, probably requests and setuptools too I don't think they are used anywhere else. I don't like to say this, but if you don't have an ubuntu box and nobody else tries it, the CI would anyways complain if something major is broken when running /test-integration

hardys replied:
Ok thanks, I'm testing locally on CentOS7 (not sure why the test-centos-integration didn't trigger?) and will try to find time to test on Ubuntu tomorrow after adjusting the package removals to match.

hardys commented:
Ok I moved away from the alias in the latest version, and instead just install a copy of the wrapper script in /usr/local/bin (which is overridable same as for kubectl)

I also removed the clients from the ubuntu pip install, but I'm not sure if we can also remove the other python libraries installed there - seems likely but can anyone confirm?

hardys commented:
/test-centos-integration

hardys commented:
Tested this again on a freshly installed RHEL 8.1 box and it works, need to verify it on CentOS7 (locally and/or in CI) and align the Ubuntu packages before removing WIP.

hardys commented:
/test-centos-integration

hardys commented:
Thanks @maelk I'm not sure how that is related to the PR atm (looks like a templating error?) but I've I'm trying to reproduce locally

hardys commented:
Ok the problem is without the RDO repos we get python-jinja 2.7.2-4 which is too old for namespace support (introduced in 2.10 - delorean-master-testing gives us 2.10-2) - looking into possible solutions (I guess ultimately this is a temporary workaround since we expect to move to Centos 8 soon)

hardys commented:
Ok I reinstated the tripleo-repos only for CentOS7 - hopefully we can soon enable CentOS8 and drop this dependency but it looks too difficult to do for CentOS7, and probably runs a big risk of breaking existing environments.

hardys commented:
/test-centos-integration

hardys commented:
/test-v1a3-integration

hardys commented:
Ok this latest version has been tested locally on CentOS7 and RHEL8 - both work (although in both cases I had to reboot before the tests would complete, will raise a bug to track that)

One thing I'd not considered is by removing the virtualbmc install on the host we lose access to the `vbmc` CLI - we should probably install a wrapper similar to `openstack` for `vbmc` which can run vbmc in the vbmc container I think

hardys commented:
/test-centos-integration

hardys commented:
/test-v1a3-integration

fmuyassarov commented on vm-setup/roles/packages_installation/defaults/main.yml:
```diff
@@ -47,19 +41,12 @@ packages:
-         - python-setuptools
          - vim-enhanced
          - wget
          - ansible
```
Do we need it here? Isn't it already installed by `centos_install_requirements.sh`, no?

hardys replied:
It could be removed but that's a different case from what I've tried to fix in this PR - I've removed things which we don't need anymore - in the case of ansible we definitely do need it, but due to the chicken/egg problem it's already installed.

For consistency I'd personally rather just leave all the packages we need in here - if the package is already installed this will do nothing but it helps see at a glance what the system dependencies are.

If there are strong opinions to the contrary we can fix it in a follow-up :)

fmuyassarov commented on vm-setup/roles/packages_installation/defaults/main.yml:
```diff
@@ -72,31 +59,22 @@ packages:
-         - python3-setuptools
          - vim-enhanced
          - wget
          - ansible
```
same here

hardys commented:
/test-centos-integration

hardys commented:
/test-v1a3-integration

maelk commented:
SOme connectivity issues on Jenkins apparently, running the tests again

maelk commented:
/test-centos-integration

maelk commented:
/test-v1a3-integration

hardys commented:
I pushed one more fix which is to allow the `vbmc` and `openstack` wrappers to work from any directory, previously the relative path of the source meant it only worked in the metal3-dev-env dir

For ease of review I added this as an extra commit

hardys commented:
/test-centos-integration

hardys commented:
/test-v1a3-integration

hardys commented:
/hold the removal of crudini on RHEL8 will cause a problem for https://github.com/openshift-metal3/dev-scripts so I need to find a solution to that before this merges

hardys commented:
https://github.com/openshift-metal3/dev-scripts/pull/933 removes the crudini dependency, and I think we can remove the similar code from metal3-dev-env as we don't need the same DNS setup for metal3-dev-env (I'll send a separate PR for that)

hardys commented:
/unhold https://github.com/openshift-metal3/dev-scripts/pull/933 merged

hardys commented:
/hold cancel

hardys commented:
@sadasu can you take another look at this please - in particular I'd like to confirm how we can ensure the Service/ServiceMonitor etc only get created for the `platform: baremetal` case where we expect the metal3 pod part to be enabled?
